Optimize Iterator.combinations

retronym opened the pull request:
Avoid a creating of temporary array just to calculate the sum
of a slice.

REPL session to show the equivalence of the new approach:
```
scala> Array(1,2,3,4).slice(1, 2).sum
res0: Int = 2

scala> Array(1,2,3,4).view(1, 2).sum
res1: Int = 2

scala> Array(1,2,3,4).slice(1, 2)
res2: Array[Int] = Array(2)

scala> Array(1,2,3,4).view(1, 2)
res3: scala.collection.mutable.IndexedSeqView[Int,Array[Int]] = SeqViewS(...)
```

Ichoran commented on src/library/scala/collection/SeqLike.scala:
```diff
@@ -232,7 +232,7 @@ trait SeqLike[+A, +Repr] extends Any with IterableLike[A, Repr] with GenSeqLike[
         _hasNext = false
       else {
-        var sum = nums.slice(idx + 1, nums.length).sum + 1
+        var sum = nums.view(idx + 1, nums.length).sum + 1
```
This is still considerably slower than a direct computation.  If it matters at all, it should be switched to an accumulator.

retronym replied:
I've pushed a new version with a plain-old loop. I've also added a unit test in a preceding commit to show the behaviour hasn't changed.

retronym committed:
Add a unit test for Seq.combinations

xuwei-k commented on src/library/scala/collection/SeqLike.scala:
```diff
@@ -9,7 +9,9 @@
 package collection
 
-import immutable.{ List, Range }
+import java.util.stream.Collectors
```
unused import?

retronym committed:
Optimize Iterator.combinations

Avoid a creating of temporary array just to calculate the sum
of a slice.

REPL session to show the equivalence of the new approach:
```
scala> Array(1,2,3,4).slice(1, 2).sum
res0: Int = 2

scala> Array(1,2,3,4).view(1, 2).sum
res1: Int = 2

scala> Array(1,2,3,4).slice(1, 2)
res2: Array[Int] = Array(2)

scala> Array(1,2,3,4).view(1, 2)
res3: scala.collection.mutable.IndexedSeqView[Int,Array[Int]] = SeqViewS(...)
```
